pkg/operator/controller/controller.go

```diff
@@ -266,11 +328,259 @@ func (r *ReconcileCDI) reconcileUpdate(logger logr.Logger, cr *cdiv1alpha1.CDI)
 		if err = r.ensureUploadProxyRouteExists(logger, cr); err != nil {
 			return reconcile.Result{}, err
 		}
+
+		if r.isUpgrading(cr) {
+			if err = r.cleanupUnusedResources(logger, cr); err != nil {
+				return reconcile.Result{}, err
+			}
+			previousVersion := cr.Status.ObservedVersion
+			cr.Status.ObservedVersion = r.namespacedArgs.DockerTag
+			if err = r.crUpdate(cdiv1alpha1.CDIPhaseDeployed, cr); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			logger.Info("Successfully finished upgrade from %v to %v and entered Deployed state", previousVersion, cr.Status.ObservedVersion)
+		}
 	}
 
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileCDI) cleanupUnusedResources(logger logr.Logger, cr *cdiv1alpha1.CDI) error {
+	//Iterate over installed resources of
+	//Deployment/CRDs/Services etc and delete all resources that
+	//do not exist in current version
+
+	targetStrategy, err := r.getAllResources(cr)
+	if err != nil {
+		return err
+	}
+
+	lo := &client.ListOptions{}
+	// maybe use different selectors?
+	lo.SetLabelSelector("cdi.kubevirt.io")
+
+	// remove unused CRDs
+	{
+		objs := &extv1beta1.CustomResourceDefinitionList{}
```

Comment:
What's with the extra nesting here?

Reply:
yeah, the scope to perform easy copy paste and to keep all the loops the same. I was hoping to find out later a way to make a method out of this loop

Reply:
fixed